py/server/deephaven/table.py

```diff
@@ -363,108 +368,306 @@ def _j_py_script_session() -> _JPythonScriptSession:
+    p_annotation.encoded_types.add(tc)
+
+
+def _parse_return_annotation(annotation: Any) -> ParsedAnnotation:
+    """ Parse an annotation in a function's signature
+
+    The return annotation is treated differently from the parameter annotations. We don't apply the same check and are
+    only interested in getting the array-like type right. Any nonsensical annotation will be treated as object type.
+    This definitely can be improved in the future.
+    """
+
+    pa = ParsedAnnotation()
+
+    t = annotation
+    pa.orig_types.add(t)
     if isinstance(annotation, _GenericAlias) and annotation.__origin__ == Union and len(annotation.__args__) == 2:
         if annotation.__args__[1] == type(None):  # noqa: E721
-            return annotation.__args__[0]
+            t = annotation.__args__[0]
         elif annotation.__args__[0] == type(None):  # noqa: E721
-            return annotation.__args__[1]
-        else:
-            return annotation
+            t = annotation.__args__[1]
+    component_char = _component_np_dtype_char(t)
+    if component_char:
+        pa.encoded_types.add("[" + component_char)
+        pa.is_array = True
     else:
-        return annotation
+        pa.encoded_types.add(_np_dtype_char(t))
+    return pa
+
+
+def _parse_numba_signature(fn: Union[numba.np.ufunc.gufunc.GUFunc, numba.np.ufunc.dufunc.DUFunc]) -> ParsedSignature:
+    """ Parse a numba function's signature"""
+    sigs = fn.types
+    if sigs:
+        p_sig = ParsedSignature(fn)
+        p_annotations = []
+        sig = sigs[0]
+        params, rt_char = sig.split("->")
+
+        ret_annotation = ParsedAnnotation()
+        ret_annotation.encoded_types.add(rt_char)
+        p_sig.params = p_annotations
+        p_sig.ret_annotation = ret_annotation
+
+        if isinstance(fn, numba.np.ufunc.dufunc.DUFunc):
+            for p in params:
+                pa = ParsedAnnotation()
+                pa.encoded_types.add(p)
+                if p in {"b", "h", "i", "l"}:
+                    pa.int_char = p
+                if p in {"f", "d"}:
+                    pa.floating_char = p
+                p_annotations.append(pa)
+        else:  # GUFunc
+            input_output_decl = fn.signature
+            input_decl, output_decl = input_output_decl.split("->")
+            input_decl = input_decl.strip("( )").split(",")
+            output_decl = output_decl.strip("( )")
+
+            for p, d in zip(params, input_decl):
+                pa = ParsedAnnotation()
+                if d:
+                    pa.encoded_types.add("[" + p)
+                    pa.is_array = True
+                else:
+                    pa.encoded_types.add(p)
+                    if p in {"b", "h", "i", "l"}:
+                        pa.int_char = p
+                    if p in {"f", "d"}:
+                        pa.floating_char = p
+                p_annotations.append(pa)
+
+            if output_decl:
+                p_sig.ret_annotation.is_array = True
```

Comment:
I think this results from `( )` indicating a scalar output.  This is not obvious from the code and should have a comment.